src/test/analyzer_test.ts

```diff
@@ -523,4 +546,211 @@ suite('Analyzer', () => {
+          new Analyzer({urlLoader: new RacyUrlLoader(contentsMap, waitFn)});
+      const promises: Promise<Document>[] = [];
+      for (let i = 0; i < 30; i++) {
+        await waitFn();
+        for (const key of contentsMap) {
+          if (Math.random() > 0.5) {
+            analyzer.analyze(key[0], key[1]);
+          }
+        }
+        promises.push(analyzer.analyze('base.html'));
+      }
+      const documents = await Promise.all(promises);
+      for (const document of documents) {
+        const imports = Array.from(document.getByKind('import'));
+        assert.deepEqual(
+            imports.map(m => m.url).sort(),
+            ['a.html', 'b.html', 'common.html', 'common.html']);
+        const docs = Array.from(document.getByKind('document'));
+        assert.deepEqual(
+            docs.map(d => d.url).sort(),
+            ['a.html', 'b.html', 'base.html', 'common.html']);
+        const refs = Array.from(document.getByKind('element-reference'));
+        assert.deepEqual(refs.map(ref => ref.tagName), ['custom-el']);
+      }
+    };
+
+    test('editor simulator of imports that import a common dep', async() => {
+      const waitTimes: number[] = [];
+      const randomWait = () => new Promise<void>((resolve) => {
+        const waitTime = Math.random() * 30;
+        waitTimes.push(waitTime);
+        setTimeout(resolve, waitTime);
+      });
+      try {
+        await editorSimulator(randomWait);
+      } catch (err) {
+        console.error('Wait times to reproduce this failure:');
+        console.error(JSON.stringify(waitTimes));
+        throw err;
+      }
+    });
+
+    /**
+     * This is a tool for reproducing and debugging a failure of the editor
+     * simulator test above, but only at the exact same commit, as it's
+     * sensitive to the order of internal operations of the analyzer. So this
+     * code with a defined list of wait times should not be checked in.
+     *
+     * It's also worth noting that this code will be dependent on many other
+     * system factors, so it's only somewhat more reproducible, and may not end
+     * up being very useful. If it isn't, we should delete it.
+     */
+    test.skip('somewhat more reproducable editor simulator', async() => {
+      // Replace waitTimes' value with the array of wait times that's logged
+      // to the console when the random editor test fails.
+      const waitTimes: number[] = [];
+
+      const reproducableWait = () => new Promise<void>((resolve) => {
+        const waitTime = waitTimes.shift();
+        if (waitTime == null) {
+          throw new Error(
+              'Was asked for more random waits than the ' +
+              'given array of wait times');
+        }
+        setTimeout(resolve, waitTime);
+      });
+      await editorSimulator(reproducableWait);
+    });
+
+    suite('deterministic tests', () => {
+      // Deterministic tests extracted from various failures of the above random
+      // test.
+
+      class Deferred<T> {
```

Comment:
thinking out loud: We really should have/import this somewhere. Would also be nice to add:

```typescript
get nodeCallback() {
  return (error, result) => {
    if (error) {
      this.reject(error)
    } else {
      this.resolve(result);
    }
  };
}
```

for manual promisification.